sqllogictest/src/runner.rs

```diff
@@ -414,102 +441,232 @@ impl<D: AsyncDB> Runner<D> {
         self.validator = validator;
     }
 
-    /// Run a single record.
-    pub async fn run_async(&mut self, record: Record) -> Result<(), TestError> {
-        tracing::info!(?record, "testing");
+    pub async fn apply_record(&mut self, record: Record) -> RecordOutput {
         match record {
-            Record::Statement { conditions, .. } if self.should_skip(&conditions) => {}
+            Record::Statement { conditions, .. } if self.should_skip(&conditions) => {
+                RecordOutput::Nothing
+            }
             Record::Statement {
                 conditions: _,
-
-                expected_error,
                 sql,
-                loc,
-                expected_count,
+
+                // compare result in run_async
+                expected_error: _,
+                expected_count: _,
+                loc: _,
             } => {
                 let sql = self.replace_keywords(sql);
                 let ret = self.db.run(&sql).await;
-                match (ret, expected_error) {
-                    (Ok(_), Some(_)) => {
-                        return Err(TestErrorKind::Ok {
-                            sql,
-                            kind: RecordKind::Statement,
+                let ret = match ret {
+                    Ok(out) => match out {
+                        DBOutput::Rows { types, rows } => RecordOutput::Query {
+                            types,
+                            rows,
+                            error: None,
+                        },
+                        DBOutput::StatementComplete(count) => {
+                            RecordOutput::Statement { count, error: None }
+                        }
+                    },
+                    Err(e) => RecordOutput::Statement {
+                        count: 0,
+                        error: Some(Arc::new(e)),
+                    },
+                };
+                if let Some(hook) = &mut self.hook {
+                    hook.on_stmt_complete(&sql).await;
+                }
```

Comment:
> Hook should be called after comparing the result.

Why? Do you mean only run hooks if query/statement succeed?

> the hook feature can be removed now as we no longer use it.

I'm not sure but it maybe useful sometimes https://github.com/risinglightdb/sqllogictest-rs/issues/88 🤪

Reply:
> only run hooks if query/statement succeed?

Previous implementation is indeed `on_query_succeed_and_pass_test` 🤪. Maybe we will also need others like `on_query_succeed_and_fail_test`... It's a little bit tricker than I thought.

Since no one is using it now, let's secretly remove it; Redesign and add back in future if needed. 😄